packages/alfa-rules/src/helpers/element-checker.ts

```diff
@@ -0,0 +1,142 @@
+import { getRole, Role } from "@siteimprove/alfa-aria";
+import { BrowserSpecific } from "@siteimprove/alfa-compatibility";
+import { Device } from "@siteimprove/alfa-device";
+import {
+  // Element,
+  getElementNamespace,
+  getInputType,
+  InputType,
+  isElement,
+  Namespace,
+  Node
+} from "@siteimprove/alfa-dom";
+import { Option } from "@siteimprove/alfa-util";
+// import { Predicate } from "@siteimprove/alfa-util";
+
+function emptyOrHas<T>(elt: T, set?: Set<T>) {
+  return set === undefined || set.has(elt);
+}
+
+export class ElementChecker {
+  constructor(namesSet?: Set<string>, inputTypeSet?: Set<InputType>) {
+    this.namesSet = namesSet;
+    this.inputTypeSet = inputTypeSet;
+  }
+
+  public withContext(context: Node): ElementCheckerWithContext {
+    return new ElementCheckerWithContext(
+      context,
+      this.namesSet,
+      this.inputTypeSet
+    );
+  }
+
+  protected namesSet?: Set<string>;
+  public withName(...names: Array<string>): ElementChecker {
+    this.namesSet = new Set(names);

```

Comment:
Storing the test in instance variables is annoying since it essentially require a new instance for every different test, and even for every "non-constant" test (_e.g._ using context).